### ISSUE TITLE
[BUG] missing notifications (duration/throttling error?) for an updated subscription

### ISSUE THREAD

**Issue by iariasleon** [telefonicaid/fiware-orion#1406]
if we create a subscription with a duration of 50s and we wait to the subscription expire, we update the same subscription with duration of 1000s, **only one notification** is sent. 

#### create subscription request
```
POST http://<host>:<port>/NGSI10/subscribeContext
Content-Type: application/json
Accept: application/json
Fiware-Service: happy_path
Fiware-ServicePath: /test
{  
     "entities": [ {                                         
            "type": "house",                                   
             "isPattern": "true",                        
             "id": "room_*"                               
      }],                                              
     "attributes": [ "temperature" ],                                
     "reference": "http://localhost:5678/notify",                                 
     "duration": "PT50S",                                 
     "notifyConditions": [ {                                           
             "type": "ONCHANGE",                          
             "condValues": ["temperature" ]                                          
       }]                                                      
       "throttling": "PT40S"
}
```

#### update subscription request
```
POST http://<host>:<port>/v1/updateContextSubscription
Content-Type: application/json
Accept: application/json
Fiware-Service: happy_path
Fiware-ServicePath: /test
{       
      "subscriptionId": "5629e7be256202e888cd1378", 
      "duration": "PT1000S"    
'}
```

**Comment by fgalan** [comment]
(From Ken's report)

New info on this issue, testing in the QA machine qa-orion-fe-01:

The test fails **only** if we use sub-cache-refresh *and* subscription-update.

sub-cache-refresh is only interesting for **active-active** and subscription-update is not really interesting at all.

NOTE:
After sub-update, the **first** notification (sometimes 2 notifications) arrives, so, it seems it is the cache-refresh that "destroys" the subscription.
A test with sub-cache-refresh-interval of 1000 secs just verified this hypothesis.

**Comment by iariasleon** [comment]
It is still failing just after 0.26.1, tested with:
```
 <version>0.26.1_20151210101150</version>
  <git_hash>1736125cdb6b4b8bffc1dd253d69fd60f0019ce2</git_hash>
```
The problem is when we try to update a subcription but it has expired previously.

**Comment by fgalan** [comment]
We have checked that expireation values is updated at DB after subcription update. However, we don't know if that update happens also in the cache: failing in doing such update could be causing the problem.

We have to check if a trace could help to check this (either an existing one or one explicitely added for the test).

**Comment by fgalan** [comment]
Some fixes have been done in the update subscription functionality regarding problems with the lost of service patch between updates (see #1693). Maybe they are related with this issue.

**Comment by fgalan** [comment]
PR https://github.com/telefonicaid/fiware-orion/pull/1998 has introduced a couple of fix in the csubs cache logic. In particular, the one related with " Fix: csub cache at mongoBackend update context subscription logic wrongly setting expiration with current time, ignoring actual expires/duration in the request" could be relate, as this bugs causes that an expiration automatically expires on update (except if the "duration" is one of the updated fields).

I'd recomment @iariasleon to re-check with develop branch once PR #1998 gets merged.

**Comment by iariasleon** [comment]
Tested, See report in https://gist.github.com/iariasleon/b19d6fc22999a13669472e1d3628998e

**Comment by fgalan** [comment]
Another bug has been discovered related with csub status sync DB->cache that causes "random" status set. To be fixes after PR #1900 merge, I'd recommend to re-check once that PR get merged.

**Comment by iariasleon** [comment]
If the throttling is greater than the cache refresh, an extra notification is sent (I think when the cache refresh is made)

throttling=5s y refresh=20s -->  sent a notification per node every 5s (OK)
throttling=40s y refresh=20s --> sent a notification per node every 40s and an extra notification

**Comment by fgalan** [comment]
A fix in cache logic has been done in PR https://github.com/telefonicaid/fiware-orion/pull/2537. Previous to that fix, subscription was deleted instead of updated if the attribute conditions vector is empty.

Not sure if this bug is related with that, but maybe it worth to test again after merging that PR.

**Comment by iariasleon** [comment]
LGTM
See https://gist.github.com/iariasleon/386fc1ddbe3e7a644633e6abd625f9a5

**Comment by fgalan** [comment]
Happy to see that this "historical issue" has been solved at the end! :)

However, given that it seems that the PR that fixed was #2537 and that PR was merged in Orion 1.4.0, I have changed the milestone.
